pkg/service/v0/groups.go

```diff
@@ -0,0 +1,321 @@
+
+// inflateGroups replaces the groups of a user with an instance that only contains the id
+func (s Service) inflateMembers(g *proto.Group) {
+	if g == nil {
+		return
+	}
+	inflated := []*proto.Account{}
+	for i := range g.Members {
+		if g.Members[i].Id != "" {
+			inflated = append(inflated, &proto.Account{Id: g.Members[i].Id})
+		} else {
+			// TODO fetch and use an id when group only has a name but no id
+			s.log.Error().Str("id", g.Id).Interface("account", g.Members[i]).Msg("resolving accounts by name is not implemented yet")
+		}
+	}
+	g.Members = inflated
+}
+
+// ListGroups implements the GroupsServiceHandler interface
+func (s Service) ListGroups(c context.Context, in *proto.ListGroupsRequest, out *proto.ListGroupsResponse) (err error) {
+
+	var query query.Query
+
+	if in.Query != "" {
+		// parse the query like an odata filter
+		var q *godata.GoDataFilterQuery
+		if q, err = godata.ParseFilterString(in.Query); err != nil {
+			s.log.Error().Err(err).Msg("could not parse query")
+			return
+		}
+
+		// convert to bleve query
+		query, err = provider.BuildBleveQuery(q)
+		if err != nil {
+			s.log.Error().Err(err).Msg("could not build bleve query")
+			return
+		}
+	} else {
+		query = bleve.NewMatchAllQuery()
+	}
+
+	s.log.Debug().Interface("query", query).Msg("using query")
+
+	searchRequest := bleve.NewSearchRequest(query)
+	var searchResult *bleve.SearchResult
+	searchResult, err = s.index.Search(searchRequest)
+	if err != nil {
+		s.log.Error().Err(err).Msg("could not execute bleve search")
+		return
+	}
+
+	s.log.Debug().Interface("result", searchResult).Msg("result")
+
+	out.Groups = make([]*proto.Group, 0)
+
+	for _, hit := range searchResult.Hits {
+
+		g := &proto.Group{}
+		if err = s.loadGroup(hit.ID, g); err != nil {
+			s.log.Error().Err(err).Str("group", hit.ID).Msg("could not load group")
+			continue
+		}
+		s.log.Debug().Interface("group", g).Msg("found group")
+
+		// TODO add accounts if requested
+		// if in.FieldMask ...
+		s.expandMembers(g)
+
+		out.Groups = append(out.Groups, g)
+	}
+
+	return
+}
+
+// GetGroup implements the GroupsServiceHandler interface
+func (s Service) GetGroup(c context.Context, in *proto.GetGroupRequest, out *proto.Group) (err error) {
+	var id string
+	if id, err = cleanupID(in.Id); err != nil {
+		return
+	}
+
+	if err = s.loadGroup(id, out); err != nil {
+		s.log.Error().Err(err).Str("id", id).Msg("could not load group")
+		// TODO we need error handling ... eg Not Found vs internal server error
+		return fmt.Errorf("internal server error")
+	}
+	s.log.Debug().Interface("account", out).Msg("found account")
+
+	// TODO add accounts if requested
+	// if in.FieldMask ...
+	s.expandMembers(out)
+
+	return
+}
+
+// CreateGroup implements the GroupsServiceHandler interface
+func (s Service) CreateGroup(c context.Context, in *proto.CreateGroupRequest, out *proto.Group) (err error) {
+	var id string
+	if in.Group == nil {
+		return fmt.Errorf("group missing")
+	}
+	if in.Group.Id == "" {
```

Comment:
If the incoming group id is NOT empty, we should check if it already exists and throw an error if it does.